core/src/main/java/org/apache/gravitino/authorization/PermissionManager.java

```diff
@@ -373,11 +377,253 @@ User revokeRolesFromUser(String metalake, List<String> roles, String user) {
+                            LOG.warn(
+                                "Securable object {} type {} doesn't exist in the role {}",
+                                object.fullName(),
+                                object.type(),
+                                role);
+                            return null;
+                          } else {
+                            // If the securable object exists, we remove the privileges of the
+                            // securable object. Remove duplicated privileges
+                            Set<Privilege> updatePrivileges = Sets.newHashSet();
+                            updatePrivileges.addAll(oldObject.privileges());
+                            privileges.forEach(updatePrivileges::remove);
+
+                            // If the object still contains privilege, we should update the object
+                            // with new privileges
+                            if (!updatePrivileges.isEmpty()) {
+                              SecurableObject newSecurableObject =
+                                  SecurableObjects.parse(
+                                      oldObject.fullName(),
+                                      oldObject.type(),
+                                      Lists.newArrayList(updatePrivileges));
+                              authorizationCallbackWrapper.setCallBack(
+                                  () ->
+                                      AuthorizationUtils.callAuthorizationPluginForMetadataObject(
+                                          metalake,
+                                          object,
+                                          authorizationPlugin -> {
+                                            authorizationPlugin.onRoleUpdated(
+                                                roleEntity,
+                                                RoleChange.updateSecurableObject(
+                                                    role, oldObject, newSecurableObject));
+                                          }));
+
+                              return newSecurableObject;
+                            } else {
+                              // If the object doesn't contain any privilege, we remove this object.
+                              authorizationCallbackWrapper.setCallBack(
+                                  () ->
+                                      AuthorizationUtils.callAuthorizationPluginForMetadataObject(
+                                          metalake,
+                                          object,
+                                          authorizationPlugin -> {
+                                            authorizationPlugin.onRoleUpdated(
+                                                roleEntity,
+                                                RoleChange.removeSecurableObject(role, oldObject));
+                                          }));
+
+                              return null;
```

Comment:
Also here, it would be better to 1) simplify the current method to split into several small methods; 2) move the callback out of transaction to avoid transaction block.

Reply:
I will split it.